### PR TITLE
feat: dynamically generate policy-related content

### PR DESCRIPTION
**Pull request by kleinfreund**
Adds a new routine before instantiating Vue which loads all available policy types from the `/policies` endpoint. This populates the new `policies`, `policiesByRoute`, and `policiesByType` states which are subsequently used to populate the sidebar menu and set-up the respective routes in the router.

BREAKING CHANGE: Due to generating the routes from the `/policies` endpoint’s response, the route paths for the Mesh Gateways (`#/mesh/all/mesh-gateway-routes` → `#/mesh/all/meshgatewayroutes`) and Mesh Gateway Routes (`#/mesh/all/mesh-gateways` → `#/mesh/all/meshgateways`) views change.

Replaces the individual policy view components (e.g. `CircuitBreakers`, etc.) with one dynamic `PolicyView` component. The individual components were mostly duplicates of each other. This made it hard to maintain because the same changes to the UI would need to be made in all components. Introducing new policies would also require changes to both Kuma and Kuma GUI. Now, depending on which route is used to access a policy view, a prop tells the component which policy type to use and looks up the respective data via the `policiesByRoute` state. Among other things, this data enables the component to construct the necessary API endpoint paths.

Changes the `DataplanePolicies` and `Meshes` components to make use of the new policy states so that they, too, work completely dynamically based on which policies are available in Kuma and which are actually set-up.

Implements #341.

Signed-off-by: Philipp Rudloff <philipp.rudloff@konghq.com>

### CODE REVIEW

**Comment by kleinfreund** [comment]
**Important**: This is blocked until the redirects for the docs website outlined in https://github.com/kumahq/kuma-gui/issues/341#issuecomment-1209014843 have been implemented.

File: src/components/DataplanePolicies/DataplanePolicies.vue
Comment (michaelbeaumont) [inline src/components/DataplanePolicies/DataplanePolicies.vue:178]: Can you add which rule is being ignore?

**Review by michaelbeaumont — APPROVED** [review]
LGTM

**Comment by michaelbeaumont** [comment]
@kleinfreund there's an error with Timeouts: https://deploy-preview-346--kuma-gui.netlify.app/#/mesh/all/timeouts

The counts are wrong, but IIRC that's an issue with the mock data, right?

File: src/components/DataplanePolicies/DataplanePolicies.vue
Comment (kleinfreund) [reply src/components/DataplanePolicies/DataplanePolicies.vue:178]: Sure, yes, I’ll ignore only the specific rule. The rule is `guard-for-in`.

**Comment by kleinfreund** [comment]
> @kleinfreund there's an error with Timeouts: https://deploy-preview-346--kuma-gui.netlify.app/#/mesh/all/timeouts
> 
> The counts are wrong, but IIRC that's an issue with the mock data, right?

@michaelbeaumont You’re correct, this is an issue with the mock data or rather the absence of it. You can observe the same issue in the current `master` branch. There just isn’t anything mocked for timeouts. I’ve added the necessary data for it now and it’s working.

**Comment by kleinfreund** [comment]
@michaelbeaumont I actually went through the entire application and found _a lot_ of inconsistencies like the one you reported in https://github.com/kumahq/kuma-gui/pull/346#issuecomment-1211199912. I assume that over time, policies, etc. got introduced and not all files were updated accordingly. A lot of places in the UI reported different counts of policies, dataplane proxies, etc. than what their respective detail views showed. I’ve tried my best to make these files consistent so that a view with 5 entries for different meshes actually showed 5 besides their entries in the sidebar.

File: src/components/DataplanePolicies/DataplanePolicies.vue
Comment (slonka) [inline src/components/DataplanePolicies/DataplanePolicies.vue:175]: ok, so this maps policies from the API to our internal representation, right? I'd have it in a separate function or maybe not mutate the data coming from the API but return a new object?

File: src/views/Policies/PolicyView.spec.ts
Comment (slonka) [inline src/views/Policies/PolicyView.spec.ts:33]: Can't we just await for some elements to be visible (like puppeteer `waitForSelector`), or [flush promises](https://v1.test-utils.vuejs.org/guides/testing-async-components.html#asynchronous-behavior-outside-of-vue)?

**Review by slonka — APPROVED** [review]
Overall seems ok, some comments + one thing that I don't like is mutating existing values, would rather have a copy of them but I'll leave it up to you to decide if you want to change that.

File: src/components/DataplanePolicies/DataplanePolicies.vue
Comment (kleinfreund) [reply src/components/DataplanePolicies/DataplanePolicies.vue:178]: There isn’t really a good reason to prevent unguarded iteration over a value that you know to be a plain object without a complex prototype chain. In my experience, I’ve never seen a stray prototype value leak into iteration and wreak havoc before. I don’t think this rule should apply in general, but tuning linting rules is a story left for another day.

File: src/views/Policies/PolicyView.spec.ts
Comment (kleinfreund) [reply src/views/Policies/PolicyView.spec.ts:33]: With Vue 3/Vue test utils 2, we can indeed simply use `flushPromises` which test utils handily provides since then. The current approach is admittedly not great, but then the upgrade is very close on the horizon. I should also just annotate that intend to use `flushPromises` which I did in a similar case elsewhere.

File: src/components/DataplanePolicies/DataplanePolicies.vue
Comment (kleinfreund) [reply src/components/DataplanePolicies/DataplanePolicies.vue]: Yes, correct. When displaying this content before, we used the original `items` data structure and the methods `getPolicyTitle` and `getPolicyLink` to extract additional data. Calling methods from templates is one of the most common pitfalls the cause additional work: when a component’s template is being re-rendered, all method calls _must_ be executed again because there is no way to tell whether its result changed. By computing this data and adding it to the `items` data structure, this is avoided.

Now as for re-using the `items` data structure as it is returned by the API call, I generally agree with the approach to mutate data. In this case, I don’t see a problem however because `items` is _completely_ local to this function. `items` _is_ completely new and there can also not be any other references to it. So mutating it directly is more direct than creating a new structure. One could make an argument about it being less resilient for certain types of refactorings, but that’s not that pressing of a matter here.

In any case, externalizing this into a separate method is definitely making this a bit cleaner. I’ve done that.

**Pull request merged**
